lib/phoenix/endpoint/instrumenter.ex

```diff
@@ -0,0 +1,138 @@
+
+      ## Examples
+
+          instrument :render_view, [view: "index.html"], fn ->
+            render conn, "index.html"
+          end
+
+      """
+      defmacro instrument(event, runtime \\ nil, fun) do
+        compile = Macro.escape(__CALLER__)
+
+        quote do
+          unquote(__MODULE__).instrument(unquote(event), unquote(compile), unquote(runtime), unquote(fun))
+        end
+      end
+
+      # For each event in any of the instrumenters, we must generate a
+      # clause of the `instrument/4` function. It'll look like this:
+      #
+      #   def instrument(:my_event, compile, runtime, fun) do
+      #     res0 = Inst0.my_event(:start, compile, runtime)
+      #     ...
+      #
+      #     start = current_time()
+      #     try do
+      #       fun.()
+      #     after
+      #       diff = ...
+      #       Inst0.my_event(:stop, diff, res0)
+      #       ...
+      #     end
+      #   end
+      #
+      for {event, instrumenters} <- events_to_instrumenters do
+        def instrument(unquote(event), var!(compile), var!(runtime), fun)
+            when is_map(var!(compile)) and is_function(fun, 0) do
+          unquote(Phoenix.Endpoint.Instrument.compile_start_callbacks(event, instrumenters))
+          start = current_time()
+          try do
+            fun.()
+          after
+            var!(diff) = time_diff(start, current_time())
+            unquote(Phoenix.Endpoint.Instrument.compile_stop_callbacks(event, instrumenters))
+          end
+        end
+      end
+
+      # Catch-all clause
+      def instrument(_event, _compile, _runtime, fun) do
+        fun.()
+      end
+
+      # Let's define the current_time/0 and time_diff/2 functions based on the
+      # existence of :erlang.monotonic_time/1.
+      @compile {:inline, current_time: 0, time_diff: 2}
+
+      if function_exported?(:erlang, :monotonic_time, 1) do
+        defp current_time, do: :erlang.monotonic_time(:micro_seconds)
+        defp time_diff(start, stop), do: stop - start
+      else
+        defp current_time, do: :os.timestamp()
+        defp time_diff(start, stop), do: :timer.now_diff(stop, start)
+      end
+    end
+  end
+
+  @doc false
+  def compile_start_callbacks(event, instrumenters) do
+    for inst <- instrumenters do
+      quote do
+        unquote(build_result_variable(inst)) =
+          unquote(inst).unquote(event)(:start, var!(compile), var!(runtime))
+      end
+    end
+  end
+
+  @doc false
+  def compile_stop_callbacks(event, instrumenters) do
+    for inst <- instrumenters do
+      quote do
+        unquote(inst).unquote(event)(:stop, var!(diff), unquote(build_result_variable(inst)))
+      end
+    end
+  end
+
+  # Given an instrumenter module, this function returns the result variable to
+  # use for that instrumenter module (the result variable is just the result
+  # returned by the start callback of an event).
+  defp build_result_variable(inst) do
+    inst
```

Comment:
Variables could be simpler. Just receive an integer based on the number of instrumenters and name them res_0, res_1, res_2 and so on. You can use `Enum.with_index` to tie the instrumenters to an index.